8324969: C2: prevent elimination of unbalanced coarsened locking regions

vnkozlov opened the pull request:
The issue is that when we do nested locks elimination we don't check that synchronized region has both `Lock` and `Unlock` nodes.
Which can happen if locks coarsening optimization eliminated pair of Unlock/Lock nodes from adjacent locking regions before we check for nested locks.

Consider this code (all locks/unlocks use the same object):
```
    lock1(obj) { // outer synchronized region
         lock2(obj) {
             // nested synchronized region 1
         } unlock2(obj);
         lock3(obj) {
             // nested synchronized region 2
         } unlock3(obj);
    } unlock1(obj);
```
If `lock3` directly follows `unlock2` (no branches or safepoints) locks coarsening optimization will remove them:

```
    lock1(obj) { // outer synchronized region
         lock2(obj) {
             // nested synchronized region 1
         };
         {
             // nested synchronized region 2
         } unlock3(obj);
    } unlock1(obj);
```

Nested locks elimination code checks only `Lock` node in one region to find if it is nested (inside other lock region which use the same object) and then eliminate it. So we end up with not eliminated `Unlock` node in second nested region.

Why we don't hit this issue before? Normally nested locks elimination executed first and only then we do locks coarsening elimination. In the example above we eliminate all nested `Lock` and `Unlock` nodes, leaving only outer `Lock` and `Unlock`.

The additional factors which leads to the failure is fully unrolled loop around nested sync regions and some allocation to trigger Escape Analysis:

```
    lock1(obj) { // outer synchronized region
         Test var = new Test(); // Triggers EA
         for (I = 0; I < 3; i++) { // small iteration number to fully unroll
             lock2(obj) {
                 // nested synchronized region 1
             } unlock2(obj);
             lock3(obj) {
                 // nested synchronized region 2
             } unlock3(obj);
         }
    } unlock1(obj);
```
Before executing Escape Analysis we do loops optimization to simplify graph: [compile.cpp#L2332](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/compile.cpp#L2332)
We also allow to fully unroll short loops (LoopOptsMaxUnroll) to remove merges from graph. It helps EA eliminate allocations.
Such unrolling creates several `Lock` and `Unlock` nodes per synchronized region. But nested locks elimination look for region with only one unique `Lock` node: [callnode.cpp#L2117](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/callnode.cpp#L2117). So we skip first iteration of nested locks elimination and execute locks coarsening optimization which eliminates all inners `Lock` and `Unlock` nodes leaving only first `Lock` and last `Unlock`. On next round of macro nodes expansion we execute nested locks elimination and remove first `Lock` node leaving `Unlock`.

The fix is to use existing list of coarsened Lock/Unlock nodes pairs to flag synchronized regions (identified by BoxLockNode nodes) as `Unbalanced` to skip any further locks elimination in these regions. 

New regression test was added. It has two `test` methods. One is based on original test from bug report. The other I wrote to show the issue clearly. Both triggers the assert.

Tested tier1-5, xcomp, stress

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324969](https://bugs.openjdk.org/browse/JDK-8324969): C2: prevent elimination of unbalanced coarsened locking regions (**Bug** - P3)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [e9f59e36](https://git.openjdk.org/jdk/pull/17697/files/e9f59e36abfa94b19bfc95849660327291a0a5aa)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [7dd6e6aa](https://git.openjdk.org/jdk/pull/17697/files/7dd6e6aa4b2774f915f4b77f7c6430f1a2f3957b)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [e5662795](https://git.openjdk.org/jdk/pull/17697/files/e566279563bfbe60b889c2be289673228b39279b)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17697/head:pull/17697` \
`$ git checkout pull/17697`

Update a local copy of the PR: \
`$ git checkout pull/17697` \
`$ git pull https://git.openjdk.org/jdk.git pull/17697/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17697`

View PR using the GUI difftool: \
`$ git pr show -t 17697`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17697.diff">https://git.openjdk.org/jdk/pull/17697.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17697#issuecomment-1925466040)

dean-long commented:
Doesn't this problem also go away if lock coarsening turns the two regions into a single region?

eme64 commented on src/hotspot/share/opto/locknode.cpp:
```diff
@@ -120,6 +128,12 @@ bool BoxLockNode::is_simple_lock_region(LockNode** unique_lock, Node* obj, Node*
+    // Unbalanced locking region.
+    // Can happen when locks coarsening optimization eliminated
+    // pair of Unlock/Lock nodes from adjacent locking regions.
+    return false;
```
Is the assumption here that each region has its own BoxLockNode, i.e. we have at most one lock and at most one unlock per box? Or can regions share a box?
If it is shared: could we have an imbalance over multiple regions that share the BoxLock, but overall we have at least one lock and one unlock, and so we pass this test?

Do I understand this right: returning false here just means that we don't do the optimization of removing the lock/unlock from the region? I think it is in `PhaseMacroExpand::mark_eliminated_box` that we would eliminate the box.

vnkozlov replied:
> Is the assumption here that each region has its own BoxLockNode, i.e. we have at most one lock and at most one unlock per box? Or can regions share a box? If it is shared: could we have an imbalance over multiple regions that share the BoxLock, but overall we have at least one lock and one unlock, and so we pass this test?

It is controlled by `EliminateNestedLocks` flag. By default the flag is ON and each synchronized region has its own BoxLockNode associated with it. The region can have one Lock node (before C2 transforms the graph) and multiple Unlock nodes (for each exit path). That is why I can't check matching numbers. I think it is possible to construct bytecode (jasm) with multiple Lock nodes for one region but in practice I never saw such case. W can have multiple Lock/Unlock nodes per region after C2 graph transformations (loop unroll/peelling) as in this case.

With `EliminateNestedLocks` flag switched off we don't do nested locks elimination and regions can share the same BoxLockNode node after Parse phase. We do only locks coarsening optimization and eliminate locks for not escaped objects.

> 
> Do I understand this right: returning false here just means that we don't do the optimization of removing the lock/unlock from the region? I think it is in `PhaseMacroExpand::mark_eliminated_box` that we would eliminate the box.

Actually the code which marks the box for nested locks is in `mark_eliminated_locking_nodes()` under `EliminateNestedLocks` flag. `is_nested_lock_region()` calls `is_simple_lock_region()` and will skip elimination for this case.

But yes, it also affects code in `mark_eliminated_box` which is called for locks with not-escaped objects. So we will not not eliminate those if we see not-balanced region (combined or not combined).

eme64 commented on test/hotspot/jtreg/compiler/escapeAnalysis/TestNestedLocksElimination.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+        return var > 0;
+    }
```
Might it make sense to add IR rules to check that we have the expected number of box, lock and unlock nodes?

vnkozlov replied:
Right.

eme64 commented:
> Doesn't this problem also go away if lock coarsening turns the two regions into a single region?

If that works that would probably be a more elegant solution. Could there be an invariant that a box always has a lock and an unlock?

vnkozlov commented:
> > Doesn't this problem also go away if lock coarsening turns the two regions into a single region?
> 
> If that works that would probably be a more elegant solution. Could there be an invariant that a box always has a lock and an unlock?

Yes, this is good suggestion. Let me investigate how to implement it. It will allow to eliminate all lock/unlock nodes in nested combined region. The only concern for me is that I have to backport the fix into all releases and if it is too complex I would prefer to go with current "band-aid" fix and have separate RFE to implement this suggestion.

vnkozlov commented:
> > > Doesn't this problem also go away if lock coarsening turns the two regions into a single region?
> > 
> > 
> > If that works that would probably be a more elegant solution. Could there be an invariant that a box always has a lock and an unlock?
> 
> Yes, this is good suggestion. Let me investigate how to implement it. It will allow to eliminate all lock/unlock nodes in nested combined region. The only concern for me is that I have to backport the fix into all releases and if it is too complex I would prefer to go with current "band-aid" fix and have separate RFE to implement this suggestion.

After looking on current code I think this will introduce more issues. In current state we have assumption that with `EliminateNestedLocks` flag ON we have only "one" Lock node per region. Depending when we start merging adjacent region it will break this  assumption.

But I share @eme64 concern that current fix may not work if coarsening eliminates an Unlock node on one of paths but left  Unlock nodes on other paths. The check will pass but we will have unbalanced lock/unlock on the first path.

I think I should just disable nested and not-escaped object lock eliminations if a coarsening elimination happens before it.

vnkozlov commented:
@eme64 and @dean-long I pushed new implementation. I used existing list of coarsened Lock/Unlock nodes pairs to flag synchronized regions (identified by BoxLockNode nodes) as unbalanced to skip any further locks elimination in these regions. Please look.

vnkozlov commented:
I did renaming `BoxLockNode::Coarsened` -> `BoxLockNode::Unbalanced`.
I plan to use it for [JDK-8322743](https://bugs.openjdk.org/browse/JDK-8322743) to mark `BoxLockNode` coming from OSR entry.

eme64 commented:
> The fix is to add check for unbalanced synchronized region (locks or unlocks) are missing.

You should probably update the PR description now.

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4948,6 +4949,34 @@ bool Compile::coarsened_locks_consistent() {
+// Mark locking regions (identified by BoxLockNode) as coarsened
+// if locks coarsening optimization removed lock/unlock from them.
+// Such regions become unbalanced and we can't execute other
+// locks elimination optimization on them.
```
The comment suggests you mark boxes as "coarsened" if ...
But really you mark them as "unbalanced", right? Because you have the condition that the locks have to already be "coarsened".

vnkozlov replied:
I updated comment

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4948,6 +4949,34 @@ bool Compile::coarsened_locks_consistent() {
+    if (size > 0) {
+      AbstractLockNode* alock = locks_list->at(0)->as_AbstractLock();
+      BoxLockNode* box = alock->box_node()->as_BoxLock();
+      if (alock->is_coarsened() && !box->is_unbalanced()) { // Not marked already
```
Is it possible that the position 0 `box` is already marked as unbalanced, but there are other `this_box` for positions `j=1..` that are not yet marked but should be?
Is it really ok that we are noch cheching all-with-all but only first-with-all?

vnkozlov replied:
You are right. I removed `!box->is_unbalanced()` check.

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4948,6 +4949,34 @@ bool Compile::coarsened_locks_consistent() {
+    }
+  }
+}
+
```
A code-style suggestion:
You use lots of nesting here, with deeper and deeper if blocks. But since you don't have any else-blocks, you could easily flatten your code by using `if (condition) { continue; }`, what I call "bailout instead of nesting". But that may be up to personal taste ;)

vnkozlov replied:
I like nesting - it clear shows conditions under which nested code is executed. And it has less lines.
I don't think this code has deep nesting. If it was more than 10 nesting I would agree to your suggestion.

eme64 reviewed (commented):
The approach looks better now :)
Though it would be good if someone who has a deeper understanding of the locking code would have a look as well.

iwanowww reviewed (approved):
Looks good.

dean-long commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4948,6 +4949,38 @@ bool Compile::coarsened_locks_consistent() {
+    }
+  }
+}
+
```
Doesn't this end up marking even the simplest lock coarsening cases as unbalanced?
Is there a reason the above check can't be done as part of coarsened_locks_consistent()?

vnkozlov replied:
> Doesn't this end up marking even the simplest lock coarsening cases as unbalanced?

Yes, the simplest lock coarsening makes locking regions as unbalanced.
See second code example in PR's description. Coarsening eliminates Unlock in regions 1 and Lock in region 2. Both regions become unbalanced.

> Is there a reason the above check can't be done as part of coarsened_locks_consistent()?

`Unbalanced` in coarsened_locks_consistent() is different from `Unbalanced` in `mark_coarsened_boxes()`
Or do you mean to recompile without lock coarsening when regions are unbalanced? But that will disable all coarsening, even valid one (for example, when no outer locking in my examples).

dean-long replied:
> > Is there a reason the above check can't be done as part of coarsened_locks_consistent()?
> 
> `Unbalanced` in coarsened_locks_consistent() is different from `Unbalanced` in `mark_coarsened_boxes()` Or do you mean to recompile without lock coarsening when regions are unbalanced? But that will disable all coarsening, even valid one (for example, when no outer locking in my examples).

If we are simply making sure that all nodes in a group have the same BoxLock, then that check could be done earlier, in coarsened_locks_consistent(), or even as the nodes are added to the list.  But I guess it's possible to get a premature answer is nodes are removed from the list.

vnkozlov replied:
I don't want to mix these two methods. They do different things:
 - `coarsened_locks_consistent()` checks that graph is correct otherwise bailout compilation
 - `mark_coarsened_boxes()` marks BoxLock nodes to prevent further locks optimizations

It could be done but it will complicate code in `coarsened_locks_consistent()` and you would still need 2 iterations:
 - first, you need to check that all `alock` on `locks_list` are marked as `coarsened` and also set flag if they reference different Boxes
 - only then you do second round over them to mark all referenced Box as unbalanced

And if we bailout from compilation Boxes marking will be useless work.

dean-long commented:
> Nested locks elimination code checks only Lock node in one region to find if it is nested (inside other lock region which use the same object) and then eliminate it. So we end up with not eliminated Unlock node in second nested region.

If lock coarsening just marked the Lock node as coarsened, but did not remove it, then Nested locks elimination would still find the Lock node and work correctly, right?  Can't we just remove coarsened locks at a later time, after locks elimination?

vnkozlov commented:
Running with test from [JDK-8322743](https://bugs.openjdk.org/browse/JDK-8322743) found that I missed `box->is_unbalanced()` check in `Lock/UnlockNode::Ideal()` where we also check for non escaped locked object.

vnkozlov commented:
> > Nested locks elimination code checks only Lock node in one region to find if it is nested (inside other lock region which use the same object) and then eliminate it. So we end up with not eliminated Unlock node in second nested region.
> 
> If lock coarsening just marked the Lock node as coarsened, but did not remove it, then Nested locks elimination would still find the Lock node and work correctly, right? Can't we just remove coarsened locks at a later time, after locks elimination?

I have old RFE for that [JDK-8268571](https://bugs.openjdk.org/browse/JDK-8268571):
```
To avoid that I suggest to execute locks coarsening after EA and nested locks optimizations.
```

The issue here is that we don't know when we done with EA and nested locks optimizations. 
Current case shows that:
 - after loop unrolling we can't do nested lock optimization because it is limited to regions with only one lock
 - we execute coarsening and it eliminated all but one lock in one regions and unlock in an other region
 - nested lock optimization now see only one lock in region and eliminate it which leads to this issue

My fix simple prevent further locks optimizations if coarsening happened.

vnkozlov commented:
Yes, we can do nested locks elimination before loop unrolling which will help in this case but may not in others.
And it is separate RFE.

vnkozlov commented:
Additional changes:
- created `ConnectionGraph::can_eliminate_lock()` to check `BoxLock` and escape status of locked object
- added check in missing places in `callnode.cpp` by calling `can_eliminate_lock()`
- renamed `mark_coarsened_boxes()` to `mark_unbalanced_boxes()`
- added more comments

eme64 reviewed (approved):
Thanks for the updates, this looks good to me now :)

shipilev commented:
Drive-by suggestion: Can we name the bug mention what is being done, rather than a symptom? Thanks!

eme64 commented:
@shipilev people have different philosophies about that, it seems. Is there a "style guiede" for that?
I prefer "what is being done" too. Makes it easier when browsing through the "git blame" later.

Suggestion: `C2: prevent elimination of unballanced coarsened lock boxes`

vnkozlov commented:
Thank you, @eme64  and @iwanowww for reviews.
@dean-long, did I answered your questions and do you have other comments?

vnkozlov commented:
> @shipilev people have different philosophies about that, it seems. Is there a "style guiede" for that? I prefer "what is being done" too. Makes it easier when browsing through the "git blame" later.
> 
> Suggestion: `C2: prevent elimination of unballanced coarsened lock boxes`

Good suggestion. I replace `boxes` with `regions`. That is what I used in code comments.

vnkozlov commented:
> My concern is that a group of unbalanced locks is added to coarsened_locks (they have different BoxLockNodes), but later some of those locks are removed by `remove_useless_coarsened_locks` or `remove_coarsened_lock`. Then when `mark_unbalanced_boxes` finally runs, it incorrectly sees the group as balanced.
> 
> It is easy to check for unbalanced in `add_coarsened_locks` as the locks are added. I think this is the most conservative approach.

I think it is premature and too conservative.

If a locking region (related lock/unlock nodes with other BoxLockNode)  is removed because it is on dead path it can affect analysis of the rest of graph. I think it is normal. Yes, group could become balanced if nodes which are left points to the same BoxLockNode. I am not sure why it is the issue - the part of graph which made it unbalanced is gone before we check consistency, BoxLock marking and locks elimination.

That is why I am doing BoxLock nodes marking after verification of `locks_list` groups consistency and before we **actually** remove marked Lock/Unlock nodes from graph. At this point graph should be correct.

@iwanowww suggested that we can additionally verify in debug VM before eliminating coarsened Lock/Unlock nodes that information in `coarsened_locks` is correct: all paths from each Lock node have corresponding Unlock node which references same object and stack slot, but may be different BoxLock. We can do it as separate RFE.

vnkozlov commented:
Looking on cases for locks coarsening [callnode.cpp#L1667](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/callnode.cpp#L1667) I don't see how it can become **balanced**.  In all cases we have separate synchronization regions (`s()`) with different BoxLock nodes. Even if one branch of code is gone the rest will still have separate regions.

I can do one more experiment. Do BoxLock marking in `add_coarsened_locks()` as you suggested and compare it in `mark_unbalanced_boxes()`.

vnkozlov commented:
@dean-long, I added verification code as I described in previous comment. Please look.
I am running tier1-5,xcopm,stress testing with it. So far did not hit new asserts.

dean-long commented:
Verification code looks correct.

vnkozlov commented:
Thank you, Dean

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4974,8 +4985,10 @@ void Compile::mark_unbalanced_boxes() {
+            assert(this_box->is_marked_unbalanced(),"inconsistency");
           }
         }
+        assert(box->is_unbalanced() == box->is_marked_unbalanced(),"inconsistency");
```
Could you even move this a scope out, i.e. a line down, so this check is even run with `alock->is_coarsened() == false`?

vnkozlov replied:
Right. It will verify that we did not miss `is_unbalanced()` check when do EA or nested elimination to avoid them.

I will also move the assert at line 4984 under `(box != this_box)` to avoid duplication of check when `box` is the same.

vnkozlov replied:
Moving assert down does not work because Nested lock elimination and EA can overwrite Coarsened status for BoxLock before we run `mark_unbalanced_boxes()`. Such BoxLock will not be marked as Unbalanced and assert will fail.
I would like to keep current assert in place but add an other assert when we change status of BoxLock. I am working on it.

vnkozlov commented:
New changes:
 - moved test to `compiler/locks` directory and renamed it because there was test there with the same name
 - introduced states for BoxLock node (and corresponding locking region)
 - added assert to verify that state change is allowed
 - added `BoxLockNode::Identity()` to correctly set state when node is "value numbered" by IGVN when `EliminateNestedLocks` flag is off.  When the flag is off C2 assumes that we have one `BoxLock` node per monitor's stack slot - we can't have merges (Phis). So I can't modify `BoxLockNode::hash()` and `cmp()`. I decide to add `Identity()` to handle that. Note, when the flag is off we have only `Regular, Coarsened and Unbalanced` states (`Local` can be set during Macro nodes elimination phase just before it is marked as `Eliminated` so it will be not visible outside small scope of code).
 - I also tested OSR case (set its BoxLock to `Unbalanced` state). I will use it for [#17331](https://github.com/openjdk/jdk/pull/17331) after I push these changes.
 - I ran tier1-4,xcomp,stress (and now running tier5-7) - no new failures. I also ran performance testing which shows no significant changes.

eme64 commented on src/hotspot/share/opto/locknode.cpp:
```diff
@@ -51,19 +51,48 @@ BoxLockNode::BoxLockNode( int slot ) : Node( Compile::current()->root() ),
+    }
+    return old_box;
+  }
+  return this;
```
For `EliminateNestedLocks` we now never have a hash, are only equal with `this`, and don't common nodes in `Ideal` either. Is that intended?

Are there any IR tests that verify that we common nodes in the cases where we expect it to common?

vnkozlov replied:
> For EliminateNestedLocks we now never have a hash, are only equal with this, and don't common nodes in Ideal either. Is that intended?

It was since JDK 8. I only corrected style there: added `{}`. And replaced `_is_eliminated` with accessor call.

It is intended for nested locks elimination - to have only one Object per locking region: [JDK-7125896@locknode.cpp](https://github.com/openjdk/jdk/commit/94927c382b40c31f4525b0c71a64e34c26e03b98#diff-d38a64764056a8888cdffdfbbd8aa31cf00ddc96040bae83112521559a7fa0a6)

> Are there any IR tests that verify that we common nodes in the cases where we expect it to common?

These optimizations were implemented before we had IR tests. We have some regression tests. During my experiments I added `_kind` to `hash()` and a lot of tests in Tier1 (not just compiler) failed when ran with `-XX:-EliminateNestedLocks`:
```
#  Internal Error (/workspace/open/src/hotspot/share/opto/parse1.cpp:2030), pid=71294, tid=28931
#  assert(!nocreate) failed: Cannot build a phi for a block already parsed.
```
So we have this verification that we common these nodes.

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -2880,6 +2881,21 @@ bool ConnectionGraph::not_global_escape(Node *n) {
+    if (EliminateNestedLocks) {
+      // We can mark whole locking region as Local only when only
+      // one object is used for locking.
+      box->set_local();
```
Could the box be `nested` at this point? If so, the assert inside `set_local` could trigger, right?

vnkozlov replied:
> Could the box be nested at this point? If so, the assert inside set_local could trigger, right?

No, it can't be nested.
Check for nesting and its elimination is done in one place from which it does not escape [macro.cpp#L2047](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/macro.cpp#L2047)
(Nested state is set only when `is_nested_lock_region()` returns true in the check).
And it will be overwritten immediately by `box_node->set_eliminated()`.

vnkozlov replied:
Also BoxLock can't be Local when we check for nesting because of `(!alock->is_non_esc_obj())` check above.

eme64 reviewed (commented):
I like the `_kind` states, with the state transitions. Makes it a bit clearer :)

Would it make sense to add the `_kind` string to the node dump?
That would possibly allow for IR matching with regex, on a few examples.
And that would make me feel a bit safer about test coverage ;)

vnkozlov reviewed (commented):
@eme64 Thank you for your comments. I hope I answered all your questions.

The only thing I may need to do is to add `const` to `mark_unbalanced_boxes()`.

eme64 reviewed (approved):
Sounds good, thanks for the explanations :)

vnkozlov commented:
Thank you, @eme64, for review.

I will merge the latest JDK sources and do additional testing with it before integration.

vnkozlov commented:
New tier1-5 testing passed.

vnkozlov commented:
/integrate
